File: qeli-win/QeliWin/Crypto/KeyDerivation.cs
using System.Security.Cryptography;
using System.Text;

namespace QeliWin.Crypto;

/// <summary>
/// HKDF-SHA256 key schedule. Direct port of the Android KeyDerivation.kt, which in
/// turn mirrors the Rust crypto/derive.rs + crypto/auth.rs. Hand-rolled HMAC
/// extract/expand to stay byte-identical with the server.
/// </summary>
public static class KeyDerivation
{
    /// <summary>SHA-256 over the in-order concatenation of the handshake records.</summary>
    public static byte[] HandshakeTranscript(IEnumerable<byte[]> records)
    {
        using var sha = SHA256.Create();
        foreach (var r in records)
            sha.TransformBlock(r, 0, r.Length, null, 0);
        sha.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
        return sha.Hash!;
    }

    /// <summary>Server-auth proof v2 (channel-bound). info = "vpn-server-auth-proof-v2" || transcript.</summary>
    public static byte[] DeriveAuthProof(byte[] staticShared, byte[] ephemeralShared, byte[] transcriptHash)
    {
        var prk = Hmac(staticShared, ephemeralShared);
        var info = Concat(Encoding.UTF8.GetBytes("vpn-server-auth-proof-v2"), transcriptHash);
        return Expand(prk, info, 32);
    }

    /// <summary>Client→server key proof. info = "vpn-client-key-proof-v1" || transcript.</summary>
    public static byte[] DeriveClientKeyProof(byte[] staticShared, byte[] ephemeralShared, byte[] transcriptHash)
    {
        var prk = Hmac(staticShared, ephemeralShared);
        var info = Concat(Encoding.UTF8.GetBytes("vpn-client-key-proof-v1"), transcriptHash);
        return Expand(prk, info, 32);
    }

    /// <summary>Returns (serverToClient, clientToServer) 32-byte ChaCha20-Poly1305 keys.</summary>
    public static (byte[] serverToClient, byte[] clientToServer) DeriveKeys(byte[] sharedSecret)
    {
        var salt = Encoding.UTF8.GetBytes("qeli-key-derivation-v1");
        var prk = Hmac(salt, sharedSecret);
        var s2c = Expand(prk, Encoding.UTF8.GetBytes("server-to-client-enc-key"), 32);
        var c2s = Expand(prk, Encoding.UTF8.GetBytes("client-to-server-enc-key"), 32);
        return (s2c, c2s);
    }

    private static byte[] Hmac(byte[] key, byte[] data)
    {
        using var mac = new HMACSHA256(key);
        return mac.ComputeHash(data);
    }

    /// <summary>RFC 5869 HKDF-Expand (single info, 1-byte counter as in the Android port).</summary>
    private static byte[] Expand(byte[] prk, byte[] info, int length)
    {
        using var mac = new HMACSHA256(prk);
        var result = new byte[length];
        var t = Array.Empty<byte>();
        int offset = 0;
        byte blockIndex = 1;
        while (offset < length)
        {
            mac.Initialize();
            var input = new byte[t.Length + info.Length + 1];
            Buffer.BlockCopy(t, 0, input, 0, t.Length);
            Buffer.BlockCopy(info, 0, input, t.Length, info.Length);
            input[^1] = blockIndex;
            t = mac.ComputeHash(input);

            int copyLen = Math.Min(t.Length, length - offset);
            Buffer.BlockCopy(t, 0, result, offset, copyLen);
            offset += copyLen;
            blockIndex++;
        }
        return result;
    }

    private static byte[] Concat(byte[] a, byte[] b)
    {
        var r = new byte[a.Length + b.Length];
        Buffer.BlockCopy(a, 0, r, 0, a.Length);
        Buffer.BlockCopy(b, 0, r, a.Length, b.Length);
        return r;
    }
}

File: qeli-ios/QeliCore/Crypto/KeyDerivation.swift
import CryptoKit
import Foundation

enum KeyDerivation {
    static func handshakeTranscript(_ records: [Data]) -> Data {
        var hash = SHA256()
        for record in records { hash.update(data: record) }
        return Data(hash.finalize())
    }

    static func serverAuthenticationProof(
        staticShared: Data,
        ephemeralShared: Data,
        transcriptHash: Data
    ) -> Data {
        let prk = hmac(key: staticShared, data: ephemeralShared)
        return expand(
            pseudorandomKey: prk,
            info: Data("vpn-server-auth-proof-v2".utf8) + transcriptHash,
            length: 32
        )
    }

    static func clientKeyProof(
        staticShared: Data,
        ephemeralShared: Data,
        transcriptHash: Data
    ) -> Data {
        let prk = hmac(key: staticShared, data: ephemeralShared)
        return expand(
            pseudorandomKey: prk,
            info: Data("vpn-client-key-proof-v1".utf8) + transcriptHash,
            length: 32
        )
    }

    static func classicKeys(sharedSecret: Data) -> (serverToClient: Data, clientToServer: Data) {
        directionalKeys(salt: "qeli-key-derivation-v1", inputKeyMaterial: sharedSecret)
    }

    static func hybridKeys(
        x25519Shared: Data,
        mlkemShared: Data
    ) -> (serverToClient: Data, clientToServer: Data) {
        directionalKeys(
            salt: "qeli-key-derivation-v2-hybrid",
            inputKeyMaterial: x25519Shared + mlkemShared
        )
    }

    static func boundClassicKeys(
        ephemeralShared: Data,
        staticShared: Data
    ) -> (serverToClient: Data, clientToServer: Data) {
        directionalKeys(
            salt: "qeli-key-derivation-v1-static-bound",
            inputKeyMaterial: ephemeralShared + staticShared
        )
    }

    static func boundHybridKeys(
        x25519Shared: Data,
        mlkemShared: Data,
        staticShared: Data
    ) -> (serverToClient: Data, clientToServer: Data) {
        directionalKeys(
            salt: "qeli-key-derivation-v2-hybrid-static-bound",
            inputKeyMaterial: x25519Shared + mlkemShared + staticShared
        )
    }

    private static func directionalKeys(
        salt: String,
        inputKeyMaterial: Data
    ) -> (serverToClient: Data, clientToServer: Data) {
        let prk = hmac(key: Data(salt.utf8), data: inputKeyMaterial)
        return (
            expand(pseudorandomKey: prk, info: Data("server-to-client-enc-key".utf8), length: 32),
            expand(pseudorandomKey: prk, info: Data("client-to-server-enc-key".utf8), length: 32)
        )
    }

    static func hmac(key: Data, data: Data) -> Data {
        Data(HMAC<SHA256>.authenticationCode(for: data, using: SymmetricKey(data: key)))
    }

    static func expand(pseudorandomKey: Data, info: Data, length: Int) -> Data {
        precondition(length > 0 && length <= 255 * SHA256.byteCount)
        let key = SymmetricKey(data: pseudorandomKey)
        var result = Data()
        var previous = Data()
        var index: UInt8 = 1
        while result.count < length {
            let input = previous + info + Data([index])
            previous = Data(HMAC<SHA256>.authenticationCode(for: input, using: key))
            result.append(previous)
            index &+= 1
        }
        return result.prefix(length)
    }
}

